src/box/box.cc

```diff
@@ -1892,6 +1911,39 @@ box_set_replication_anon(void)
 	guard.is_active = false;
 }
 
+/**
+ * Set a new cluster name directly, bypassing all checks like whether the
+ * instance is writable. It makes the function usable by bootstrap master when
+ * it is read-only but has to make the first registration.
+ */
+static void
+box_set_cluster_name_direct(const struct tt_hostname *name)
+{
+	int rc;
+	if (tt_hostname_is_nil(name)) {
+		rc = boxk(IPROTO_DELETE, BOX_SCHEMA_ID, "[%s]", "cluster_name");
+	} else {
+		rc = boxk(IPROTO_REPLACE, BOX_SCHEMA_ID, "[%s%s]",
+			  "cluster_name", tt_hostname_str(name));
+	}
+	if (rc != 0)
+		diag_raise();
+}
+
+void
+box_set_cluster_name(void)
+{
+	struct tt_hostname name;
+	box_check_cluster_name(&name);
+	/* Nil means the config doesn't care, allows to use any name. */
+	if (tt_hostname_is_nil(&name))
+		return;
+	if (tt_hostname_are_eq(&CLUSTER_NAME, &name))
+		return;
+	box_check_writable_xc();
+	box_set_cluster_name_direct(&name);
```

Comment:
* *"The fact that box.cfg may actually do DDL under the hood looks rather unexpected."* - I wouldn't even call this DDL. `_schema` is an internal space (like you said). It just happened to be a space really, but its keys could easily be stored in WAL as special records, like Raft does. So I don't see it as a problem. While we are here - does turning an anon replica to a normal one count as DDL? The replica asks the master to insert its record into `_cluster`. I would not call such manipulations with system spaces DDL.

* *"cluster_name may not be changed on initial configuration, only dynamically, after box init"* - this is known and approved by you, Sergey Os., Sergey P., and Mons. We discussed it in the chat. There were options how to make the initial `box.cfg` able to change all the names, but those approaches looked too complicated or too confusing to bother, given the purpose of renames (do it one time on upgrade).

* *"force_recovery may actually affect other subsystems so setting it to change cluster_name may have unwanted side effects."* - yes, I am aware. I don't want to add new `box.cfg` option each time when I need to ignore specific errors concerning WAL. `box.cfg` is already bloated enough.

About `box.ctl`. It would be 3 new functions, one for each name. This would bloat the ctl API and would make an impression like the renames are legal. But they are not. They are not supposed to be easy to do. Rename is only allowed now to allow people migrate from < 3.0 versions.

Another problem is that if you change cluster or replicaset name on a master, then the replicas have to allow that rename. Currently you can set them to `force_recovery = true`, otherwise they will reject the rename. If we would only have a ctl function, then how should replicas understand that the rename is allowed? Have both `force_recovery` and ctl functions?

Another problem is just local, not about replicas. The names are stored in `box.cfg`. Like you can get them as `box.cfg.cluster_name`, `box.cfg.replicaset_name` if you did configure them on this instance. If you change a name via a ctl function, the config will become desynced with the actually stored name. Again that means you wouldn't be able to avoid `box.cfg` manipulations. Firstly somehow allow the rename (`force_recovery`?), secondly - change the config (`box.cfg{name = new_name}`?).

On the summary, it doesn't seem easier to use with some new ctl functions. You still have to bother with `box.cfg`. Or did you have some specific ideas how to workaround those issues?

Reply:
> * _"The fact that box.cfg may actually do DDL under the hood looks rather unexpected."_ - I wouldn't even call this DDL. `_schema` is an internal space (like you said). It just happened to be a space really, but its keys could easily be stored in WAL as special records, like Raft does. So I don't see it as a problem. While we are here - does turning an anon replica to a normal one count as DDL? The replica asks the master to insert its record into `_cluster`. I would not call such manipulations with system spaces DDL.

Fair enough.

> * _"cluster_name may not be changed on initial configuration, only dynamically, after box init"_ - this is known and approved by you, Sergey Os., Sergey P., and Mons. We discussed it in the chat. There were options how to make the initial `box.cfg` able to change all the names, but those approaches looked too complicated or too confusing to bother, given the purpose of renames (do it one time on upgrade).

Sorry, I didn't really put much thought in it back then :(

> * _"force_recovery may actually affect other subsystems so setting it to change cluster_name may have unwanted side effects."_ - yes, I am aware. I don't want to add new `box.cfg` option each time when I need to ignore specific errors concerning WAL. `box.cfg` is already bloated enough.

Adding a new `force_recovery` option isn't a good idea, I agree.

> About `box.ctl`. It would be 3 new functions, one for each name. This would bloat the ctl API and would make an impression like the renames are legal. But they are not. They are not supposed to be easy to do.

Technically, we could add one function that would take a table containing any subset of all the available names. We could also add an option table to this function and allow rename only if the `force` flag is set.

> Rename is only allowed now to allow people migrate from < 3.0 versions.

Are we planning to forbid renaming eventually? If so, that is if this option is required only for upgrade, maybe we could pass the names to `box.schema.upgrade()` somehow and don't allow to change them with `box.cfg` at all?

> Another problem is that if you change cluster or replicaset name on a master, then the replicas have to allow that rename. Currently you can set them to `force_recovery = true`, otherwise they will reject the rename. If we would only have a ctl function, then how should replicas understand that the rename is allowed? Have both `force_recovery` and ctl functions?

I think that replicas should successfully apply renaming received from the master even if `box.cfg.force_recovery` is false, otherwise upgrade is going to become painful (the admin would have to set `force_recover` on all replicas). I think it's okay to unconditionally trust changes received via a replication channel, at least in this particular case.

> Another problem is just local, not about replicas. The names are stored in `box.cfg`. Like you can get them as `box.cfg.cluster_name`, `box.cfg.replicaset_name` if you did configure them on this instance. If you change a name via a ctl function, the config will become desynced with the actually stored name. Again that means you wouldn't be able to avoid `box.cfg` manipulations. Firstly somehow allow the rename (`force_recovery`?), secondly - change the config (`box.cfg{name = new_name}`?).

That's true, but the same problem is present in the current approach, but at the replicas' side. Renames received via a replication channel won't be reflected in replica's `box.cfg` AFAIU. We don't care about that so I guess we shouldn't care about out-of-sync `box.cfg` on the master, either. After all, `box.cfg` isn't the ultimate source of truth - `box.info` is.

Reply:
> Are we planning to forbid renaming eventually? If so, that is if this option is required only for upgrade, maybe we could pass the names to box.schema.upgrade() somehow and don't allow to change them with box.cfg at all?

I suppose yes, rename ban is the intention for the future. Although PMs can always change their mind.

I don't like here the duplication of upgrade and cfg options. Not that it is impossible to do, but it just looks more complicated than the current scheme.

> I think that replicas should successfully apply renaming received from the master even if box.cfg.force_recovery is false, otherwise upgrade is going to become painful (the admin would have to set force_recover on all replicas)

This is a grey zone then. If a replica does rebootstrap, should it still care about the names mismatch? Technically, rebootstrap is like a new replica. But we tell people that it is the same replica, just fetching a new snapshot. If the names don't need to be validated, then what is their purpose besides showing them in `box.info`? I had an impression that one of their usages is to cover all usecases of UUIDs, including validation of the replicas being located in the correct cluster/replicaset. Without names validation this protection won't work.

> That's true, but the same problem is present in the current approach, but at the replicas' side.

Nope. On the replicas you have to set `force_recovery` first. It means their config won't be broken. I.e., if after the new name was delivered everywhere and they would do `box.cfg{totable(box.cfg)}` or something like that (re-apply the same config), this `box.cfg` will succeed. On the other hand, if you only have ctl, then re-applying the same config won't work - the name is mismatching and it is not allowed (`force_recovery` is not set). (I know that we can't reapply the same config, `box.cfg` will just skip the unchanged params, but imagine it doesn't to see the point. Or that the node was restarted after receiving the new name.)

> After all, box.cfg isn't the ultimate source of truth - box.info is.

I would reformulate - `box.info` is the actual state, `box.cfg` is the desirable state (not for all params, but for those which are reflected in `box.info` and set in `box.cfg`, like `replication_anon` in cfg and `box.info.id == 0` in info).

If you think of that this way, then you might see why we need validation. It might happen that the actual state just is not compatible with the desired one. For instance, `replication_anon = false` might fail non-critically, with a Lua error thrown. Or `replication_uuid = <value>` might mismatch with what is received from the master, then `box.cfg` will fail critically. Won't even boot.

Reply:
> I suppose yes, rename ban is the intention for the future. Although PMs can always change their mind.
> 
> I don't like here the duplication of upgrade and cfg options. Not that it is impossible to do, but it just looks more complicated than the current scheme.

I wouldn't agree that it'd be more complicated:
 - From the user point of view it should be easier because you'd just pass the names to `box.schema.upgrade()` called on a master instance and wouldn't have to set/unset `force_recovery` on all instances.
 - From the implementation point of view it shouldn't be difficult either. For example, we could simply ignore all checks if the `dd_version` is less than 3.0.

The only downside I see is that you can't change names afterwards, which may be required in future.

Another idea: What it we allow to set names via `box.cfg` but skip all checks until `dd_version_id` is bumped to 3.0?

> This is a grey zone then. If a replica does rebootstrap, should it still care about the names mismatch? Technically, rebootstrap is like a new replica. But we tell people that it is the same replica, just fetching a new snapshot. If the names don't need to be validated, then what is their purpose besides showing them in `box.info`? I had an impression that one of their usages is to cover all usecases of UUIDs, including validation of the replicas being located in the correct cluster/replicaset. Without names validation this protection won't work.

AFAIU a replica should check the names only on connect. If they match it should accept all changes from the master. I don't see why this wouldn't work. In your example with rebootstrap, the replica would fail to connect in the first place if the names are different.

> Nope. On the replicas you have to set `force_recovery` first. It means their config won't be broken. I.e., if after the new name was delivered everywhere and they would do `box.cfg{totable(box.cfg)}` or something like that (re-apply the same config), this `box.cfg` will succeed.

Reapplying the same config wouldn't work AFAIU. The user'd have to manually set the new names because the original config should contain old names for the replica to start.

> On the other hand, if you only have ctl, then re-applying the same config won't work - the name is mismatching and it is not allowed (`force_recovery` is not set).

Technically, `box.ctl` could update the `box.cfg` config (just like the `log` module does). Still, this wouldn't solve the problem on replicas - the config'd have to be updated manually.

Reply:
> Another idea: What it we allow to set names via box.cfg but skip all checks until dd_version_id is bumped to 3.0?

This might be possible. I can do that, but you need to talk to a PM first to get an approval.

> AFAIU a replica should check the names only on connect. If they match it should accept all changes from the master. I don't see why this wouldn't work.

If you changed the name on master and then the replica disconnected suddenly, then it won't ever reconnect until the master changes the name back.

> In your example with rebootstrap, the replica would fail to connect in the first place if the names are different.

You can change the name in the replica config to the one used on the master and it will connect. You can also start the replica with empty `replication` and then connect with `force_recovery` to fetch the new name. You can also set `instance_name` or any other name to `nil/box.NULL` and it will connect to anything (unless the replica has already persisted a name different from the master - then you need to do one of the previous ways).

> box.ctl could update the box.cfg config

It is not about displaying the `box.cfg.<key>` values. The actual config is stored outside of box by the user. You can't change it. It might arrive from the network or from a vshard config and the entire config will re-applied. The updated keys will change.

So even if you update `box.cfg.instance_name` to a new value, then a next complete re-cfg coming from a user-owned config storage will revert it back to the old value and break.

The name update via cfg forces people to do it via their configs without even writing any code, not via ctl manually.

Reply:
Before I redo the patchset third time, I would like this comment either to be marked resolved or to see more reasons why I should change anything. For behaviour and API changes which we discussed but you still don't agree with, you need to talk to a PM, for example to @Mons.